File: archive/php/muse/data_structures/SinglyLinkedList.php
<?php

declare(strict_types=1);

namespace muse\data_structures;

require_once "muse/data_structures/List.php";

use muse\data_structures\Lister;
use muse\data_structures\SinglyLinkedList\Node;

class SinglyLinkedList implements Lister
{
    private $head = null;
    private $length = 0;

    public function __construct()
    {
    }

    public function size(): int
    {
        return $this->length;
    }

    public function isEmpty(): bool
    {
        return $this->length == 0;
    }

    public function get(int $index): int
    {
        if ($index < 0 || $index >= $this->length) {
            throw new RuntimeException("[PANIC - IndexOutOfBounds]");
        }

        $cursor = $this->head;

        for ($i = 0; $i < $index; $i++) {
            $cursor = $cursor->next;
        }

        return $cursor->data;
    }

    public function set(int $index, int $element): void
    {
        if ($index < 0 || $index >= $this->length) {
            throw new RuntimeException("[PANIC - IndexOutOfBounds]");
        }

        $cursor = $this->head;

        for ($i = 0; $i < $index; $i++) {
            $cursor = $cursor->next;
        }

        $cursor->data = $element;
    }

    public function insert(int $index, int $element): void
    {
        if ($index < 0 || $index > $this->length) {
            throw new RuntimeException("[PANIC - IndexOutOfBounds]");
        }

        $node = new Node($element);

        if ($index == 0) {
            if ($this->length != 0) {
                $node->next = $this->head;
            }
            $this->head = $node;
        } else {
            $cursor = $this->head;
            for ($i = 0, $bound = $index - 1; $i < $bound; $i++) {
                $cursor = $cursor->next;
            }
            $node->next = $cursor->next;
            $cursor->next = $node;
        }

        $this->length++;
    }

    public function remove(int $index): void
    {
        if ($index < 0 || $index >= $this->length) {
            throw new RuntimeException("[PANIC - IndexOutOfBounds]");
        }

        $target = null;

        if ($index == 0) {
            $target = $this->head;
            if ($this->length == 1) {
                $this->head = null;
            } else {
                $this->head = $target->next;
            }
        } else {
            $cursor = $this->head;
            for ($i = 0, $bound = $index - 1; $i < $bound; $i++) {
                $cursor = $cursor->next;
            }
            $target = $cursor->next;
            $cursor->next = $target->next;
        }

        $target->data = null;

        $this->length--;
    }

    public function front(): int
    {
        return $this->get(0);
    }

    public function back(): int
    {
        return $this->get($this->length - 1);
    }

    public function prepend(int $element): void
    {
        $this->insert(0, $element);
    }

    public function append(int $element): void
    {
        $this->insert($this->length, $element);
    }

    public function poll(): void
    {
        $this->remove(0);
    }

    public function eject(): void
    {
        $this->remove($this->length - 1);
    }
}

namespace muse\data_structures\SinglyLinkedList;

class Node
{
    public $data;
    public $next = null;

    public function __construct(int $element)
    {
        $this->data = $element;
    }
}

File: archive/ruby/muse/data_structures/SinglyLinkedList.rb
module SinglyLinkedList
  class Node
    attr_accessor :data, :next

    def initialize(element)
      @data = element
      @next = nil
    end
  end

  class SinglyLinkedList
    def initialize
      @head = nil
      @length = 0
    end

    def size
      @length
    end

    def isEmpty
      @length.zero?
    end

    def get(index)
      raise "[PANIC - IndexOutOfBounds]" if index.negative? || index >= @length

      cursor = @head

      index.times do
        cursor = cursor.next
      end

      cursor.data
    end

    def set(index, element)
      raise "[PANIC - IndexOutOfBounds]" if index.negative? || index >= @length

      cursor = @head

      index.times do
        cursor = cursor.next
      end

      cursor.data = element
    end

    def insert(index, element)
      raise "[PANIC - IndexOutOfBounds]" if index.negative? || index > @length

      node = Node.new(element)

      if index.zero?
        node.next = @head if @length != 0
        @head = node
      else
        cursor = @head
        (index - 1).times do
          cursor = cursor.next
        end
        node.next = cursor.next
        cursor.next = node
      end

      @length += 1
    end

    def remove(index)
      raise "[PANIC - IndexOutOfBounds]" if index.negative? || index >= @length

      target = nil

      if index.zero?
        target = @head
        @head = if @length == 1
          nil
        else
          target.next
        end
      else
        cursor = @head
        (index - 1).times do
          cursor = cursor.next
        end
        target = cursor.next
        cursor.next = target.next
      end

      target.data = nil

      @length -= 1
    end

    def front
      get(0)
    end

    def back
      get(@length - 1)
    end

    def prepend(element)
      insert(0, element)
    end

    def append(element)
      insert(@length, element)
    end

    def poll
      remove(0)
    end

    def eject
      remove(@length - 1)
    end
  end
end
